Workiva/over_react
# FED-2432: Don't lint required props that are set in factory

## Motivation
There are some common use-cases where consumers declare BuilderOnlyUiFactory functions that are to be used as the main entrypoint of their components, and assign required props within them.

For example, the custom `FooComponents.content` factory:
```dart
UiFactory<FooProps> Foo = ...;
class FooProps = UiProps with FluxUiPropsMixin<...>, ...;

class FooComponents {
  ...
  FooProps content() => Foo()
    ..store = store
    ..actions = actions;
}
```

Currently, the `missing_required_prop` lint occurs when using that factory, even though it sets required props:
```dart
void renderApp(FooComponents components) {
  react_dom.render(components.content()(), mountNode);
  //               ^^^^^^^^^^^^^^^^^^^^
  // warning: over_react_late_required_prop 
  //   Missing required late prop 'actions' from 'FluxUiPropsMixin'
  //   Missing required late prop 'store' from 'FluxUiPropsMixin'
}
```

As opposed to making all consumers of those types of factories ignore errors, or some other solution involving annotation, we want to not lint when we can determine which props are set in the factory.

## Changes
- Add `getPropsSetByFactory` utility to look up props set in factories, add tests
- Update `missing_required_prop` lint
    - Take into account props set by factories, add tests
    - Update `// debug: over_react_required_props` information with props set by factory, and whether props set by factory were computed
- Add playground example
- Update function Expando-based memoization to work with nullable return values (this wasn't 100% necessary, but cleaned things up a fair bit)
   - Replace `_memoizeWithExpando` usages with with `memoizeWithWeakMap`
   - Refactor `WeakMap` to support nullable values, add tests
   - Add `memoizeWithWeakMap` tests


#### Release Notes
  

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        - CI passes
        - Test in playground example that:
            - The diagnostic no longer lints for props set in factories
            - Add `// debug: over_react_required_props` and verify that _propsSetByFactory needed to be computed_ is true **only** for usages where either:
               - props were set in the factory
               - there are missing required props that resulted in a lint

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
